### PR TITLE
Link to previous blames in file blame page

### LINKED ISSUE #15109: feature request: previous blames
githib allows you to navigate the history of changes on a per-line basis:

image

gitea (1.13.2, at least) doesn't seem to allow this. you can view the blame for a file, and you can click to see the commit for each line-group, but there's no way (that i can see) to navigate to the previous commit for that line.

### PR DESCRIPTION
**Pull request by noerw**
Adds a link to each blame hunk, to view the blame of an earlier version of the file, similar to GitHub. Also refactors the blame render from fmtstring based to template based.

This supersedes #15171: I copied that PR, resolved conflicts, and adressed my code review concerns, as the original author seems unresponsive.

old demo gif:

Fixes #15109
Closes #15171

### CODE REVIEW

File: routers/web/repo/blame.go [hunk 1]
```diff
@@ -164,6 +178,17 @@ func RefBlame(ctx *context.Context) {
...
+		previousCommit, err := commit.Parent(0)
+		if err == nil {
+			// Verify the commit
+			if haz, _ := commit.HasPreviousCommit(previousCommit.ID); haz {
+				if haz1, _ := previousCommit.HasFile(ctx.Repo.TreePath); haz1 {
+					previousCommits[commit.ID.String()] = previousCommit.ID.String()
+				}
+			}
+		}
+
```
Comment (zeripath) [inline routers/web/repo/blame.go]: Argh!

commit.HasPreviousCommit is not free and is potentially computationally very expensive. It involves a call to `git merge-base --is-ancestor`. The results of this need to be cached at least within the loop. 

Further I don't think:

`haz, _ := commit.HasPreviousCommit(previousCommit.ID)` can ever be false as it is a 0-level parent by line 182: `previousCommit, err := commit.Parent(0)`.

`haz1, _ := previousCommit.HasFile(ctx.Repo.TreePath); haz1` - is also quite expensive and should be cached.

Comment (noerw) [reply routers/web/repo/blame.go]: Oh ok. ~~Is there a common cache facility to use for this? Or should I just do an adhoc `map[string]bool` cache?~~

I assume you mean to implement something analog to `LastCommitCache` like `CommitHasPathCache`?

Comment (zeripath) [reply routers/web/repo/blame.go]: probably just use a just out-of-loop cache for the moment. In this case I don't think there's likely to be benefit pushing up to `modules/cache` or using an `*lru.TwoQueue`/`*lru.Cache`  directly.

We should consider whether to cache the results of the internal `git.Commit.GetFiles` in modules/cache.Cache but at present the default cache is unlimited unless we merge my PR.

Comment (zeripath) [reply routers/web/repo/blame.go]:
```diff
diff --git a/routers/web/repo/blame.go b/routers/web/repo/blame.go
index 9efb868c8..79612f0d5 100644
--- a/routers/web/repo/blame.go
+++ b/routers/web/repo/blame.go
@@ -161,6 +161,8 @@ func RefBlame(ctx *context.Context) {
 	commitNames := make(map[string]models.UserCommit)
 	previousCommits := make(map[string]string)
 	commits := list.New()
+	commitCache := map[string]*git.Commit{}
+	commitCache[commitID] = commit
 
 	for _, part := range blameParts {
 		sha := part.Sha
@@ -168,21 +170,30 @@ func RefBlame(ctx *context.Context) {
 			continue
 		}
 
-		commit, err := ctx.Repo.GitRepo.GetCommit(sha)
-		if err != nil {
-			if git.IsErrNotExist(err) {
-				ctx.NotFound("Repo.GitRepo.GetCommit", err)
-			} else {
-				ctx.ServerError("Repo.GitRepo.GetCommit", err)
+		commit, ok := commitCache[sha]
+		if !ok {
+			commit, err = ctx.Repo.GitRepo.GetCommit(sha)
+			if err != nil {
+				if git.IsErrNotExist(err) {
+					ctx.NotFound("Repo.GitRepo.GetCommit", err)
+				} else {
+					ctx.ServerError("Repo.GitRepo.GetCommit", err)
+				}
+				return
 			}
-			return
+			commitCache[sha] = commit
 		}
 
-		// Get previous closest commit sha from the commit
-		previousCommit, err := commit.Parent(0)
-		if err == nil {
-			// Verify the commit
-			if haz, _ := commit.HasPreviousCommit(previousCommit.ID); haz {
+		if len(commit.Parents) > 0 {
+			psha := commit.Parents[0]
+			previousCommit, ok := commitCache[psha.String()]
+			if !ok {
+				previousCommit, _ = commit.Parent(0)
+				if previousCommit != nil {
+					commitCache[psha.String()] = previousCommit
+				}
+			}
+			if previousCommit != nil {
 				if haz1, _ := previousCommit.HasFile(ctx.Repo.TreePath); haz1 {
 					previousCommits[commit.ID.String()] = previousCommit.ID.String()
 				}
```

**Comment by zeripath** [comment]
@noerw can you explain what this previous commit thing is supposed to be doing? I'm a little confused by the algorithm.

File: routers/web/repo/blame.go (hunk 1 above)
Comment (noerw) [reply routers/web/repo/blame.go]: thanks, done

**Comment by noerw** [comment]
> @noerw can you explain what this previous commit thing is supposed to be doing? I'm a little confused by the algorithm.

@zeripath The idea is to replicate behaviour as in github, where the previous-blame button doesn't link to the commit the blamePart is related to, but its parent. This behaviour is nice, as it avoids to click twice through previous commit links..

We only want to show this link, when the file actually exists in the parent commit, which needs to be validated as the parent commit may be older than the first commit for this file. We could of course just display the link, which then leads to a 404, if the perf trade of is not worth it..

**Review by zeripath — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Pull request merged**